Use Hamcrest in unit tests, it gives better diagnostics

proski opened the pull request:
```
*  Use Hamcrest in unit tests, it gives better diagnostics
   
   Add explicit JUnit dependency in pom.xml and make sure that Hamcrest
   stands above JUnit, as suggested in Hamcrest documentation.
```

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/AdditionalParameterRegExTest.java:
```diff
@@ -12,65 +17,65 @@
-    assert ("".equals(test1.getValue()));
+    assertNotNull(test1);
+    assertEquals("the_blank_parameteR", test1.getKey());
+    assertEquals("", test1.getValue());
```
Isn't there some assertBlank? I'm sure there is a hamcrest matcher that does this

proski replied:
`assertThat(test1.getValue(), isEmptyString());`

jakub-bochenski commented:
This change is an improvement over the previous state. What would be even better is converting the whole thing to hamcrest/assertJ, then instead of this:
```
    assertNotNull(test1);
    assertEquals("customer", test1.getKey());
    assertEquals("Nickesocke", test1.getValue());
```
You would just get something like this:
```
 assertThat(test1, both(
       hasKey("customer"))
       .and(hasValue("Nickesocke"))
  ));
```

proski committed:
Use Hamcrest in unit tests, it gives better diagnostics

Add explicit JUnit dependency in pom.xml and make sure that Hamcrest
stands above JUnit, as suggested in Hamcrest documentation.

Rename "testN" variables to "entry", as they are instances of
SimpleEntry.

proski commented:
Converted to Hamcrest 1.3.

proski commented:
Update to Hamcrest 2.1 after some consideration. It has much better support for maps. I don't see other plugins or Jenkins itself using Hamcrest 2.1, but I think it's OK. It's our internal test dependency.

I considered assertJ, it's nice, but we should also consider that it's less popular.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/AdditionalParameterRegExTest.java:
```diff
@@ -12,65 +15,65 @@
-    assert (multipleParameters.get("param2").equals("something special"));
-    assert (multipleParameters.get("param3").equals(""));
+    assertThat(multipleParameters, is(aMapWithSize(3)));
+    assertThat(multipleParameters.get("param1"), is("nothing"));
```
How about `multipleParameters, hasEntry("param1","nothing")` instead?

proski replied:
Done, good idea.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/stash/AdditionalParameterRegExTest.java:
```diff
@@ -12,65 +15,65 @@
     AbstractMap.SimpleEntry<String, String> test4 = StashRepository.getParameter("p:=nothing");
-    assert (test4 == null);
+    assertThat(test4, is(nullValue()));
     AbstractMap.SimpleEntry<String, String> test5 = StashRepository.getParameter("=nothing");
```
Can you inline the local variables? `test5` is an awful name

proski replied:
I agree about awful names. I just used `entry` everywhere. I'm not sure about inlining, it would mix test setup and test verification into the same expression. I can do it is you insist, but I'd rather not.

jakub-bochenski replied:
My point is: if you can't give the variable a good name, don't extract it. `Entry` doesn't bring any more information (it is a bit better than `test5`).

```
    assertThat(
            StashRepository.getParameter("=nothing"),
            is(nullValue())
   );
```
I think above is easier to read, as you don't need to keep the extra `entry` in your head. There is a also the in-elegance of re-assigning the local variable. I would prefer local variables to be final.

It's not a big problem though, the tests still look better overall after this change

proski replied:
I've inlined the variables wherever practical. `singleParameters` and `multipleParameters` need multiple checks, so they remain named variables.

jakub-bochenski replied:
Thanks
The previous version was good enough for me, I'm just waiting for possible input from @jimklimov

proski commented:
Replaced deprecated `isEmptyString()` with `is(emptyString())`
